### PR TITLE
feat: add kurl command line for disk space check

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:

this commits adds a new kurl command to check the amount of available space, by storage class, in the cluster.

usage example:
```
kurl cluster free-disk-space [--storageclass openebs] [--openebs-node node] [--openebs-image eeacms/rsync:2.3]
```

output examples:

```
$ kubectl get storageclasses
NAME                PROVISIONER                     RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
default (default)   rook-ceph.rbd.csi.ceph.com      Delete          Immediate              true                   69m
openebs             openebs.io/local                Delete          WaitForFirstConsumer   false                  64m
rook-cephfs         rook-ceph.cephfs.csi.ceph.com   Delete          Immediate              true                   69m

$ /tmp/kurl cluster free-disk-space --storageclass openebs
62421426176     ip-172-31-60-177
62894407680     ip-172-31-80-103

$ # if no --storageclass flag is provided it looks for the default one.
$ /tmp/kurl cluster free-disk-space
85880889344
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Does this PR require documentation?
NONE

### CODE REVIEW

**Review by laverya — DISMISSED** [review]

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:26]: Would be possible doc the `NewClusterFreeDiskSpaceCmd ` func?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:39]: Could we properly encapsulate the methods into validate() and then run()?

File: pkg/cli/commands.go
Comment (camilamacedo86) [inline pkg/cli/commands.go:24]: The name/UX here does not seems intuitive for me:

> kurl cluster free-disk-space [--storageclass openebs] [--openebs-node node] [--openebs-image eeacms/rsync:2.3]

Will it ensure that we have a free disk space or is it checking validating that we have a free space?

Therefore, I understand that we have many checks in commands such this one.
So, we could looking for in the future how we could better organize and etc the subcommands.

However, for now could we call this one check-free-disk-space? Is that OK? WDYT?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:37]: Why are using `SilenceUsage `?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:38]: Could we add an Example desc with example over how to call it? Also, could we add an Long desc describing when/why/it is required useful?  

See an example
https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/init.go#L64-L76

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:122]: Why the value is `eeacms/rsync:2.3`?
What happens if we inform an invalid option?
Should we not validate the inputs? 

Also, can we run the command without inform this flags? If not, then we need mark then as required, see `MarkFlagRequired("name")`

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Could we try before merge: 

- ensure that required flags are marked as required
- invalid flag inputs will return an error
- we have an example desc so that is possible to know looking in the code how it is used
- we have a long desc describing when/why/where it is useful / required?
- we have any exported const or func properly comment so that new code would not fail in lint checks when we be able to enable them.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:28]: fixed.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:41]: Could you please point me to an example of this ?

File: pkg/cli/commands.go
Comment (ricardomaraschini) [reply pkg/cli/commands.go]: fixed.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:39]: Copy and Pasta. What are you worried about ?

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:124]: `eeacms/rsync:2.3` is the default image used by `pvmigrate`, that is why it is here. if an invalid input is provided the command will fail and an error will be reported back. all flags are optional.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:126]: Added in the `Example` property.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:41]: I have added and example to the `Example` property. Is that enough ?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:74]: If that fails why what we output here?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:80]: If that fails why what we output here?

Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:80]: this is what have been logged by the "freeSpaceChecker".

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:74]: this is what have been logged by the "freeSpaceChecker".

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:86]: Question how is the expected result? 

value
value
value 
error? 

Should we not show the total? Could you please add an screenshot of your test for we see how that would to be?

Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:86]: No, we must show per node (for openebs). There is an example in the description of this PR.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:42]: I thought in something like: 


```go
// NewRookFreeDiskSpaceGetter returns a command that is capable of reporting back the amount of free space in the cluster
// for a provided storage class.
func NewClusterFreeDiskSpaceCmd(cli CLI) *cobra.Command {
	var forStorageClass, openEBSImage, openEBSNode string
	logger = log.New(logbuf, "", 0)

	cmd := &cobra.Command{
		Use:          "free-disk-space",
		Short:        "Returns the amount of free disk space (in bytes) for a given storage class",
		SilenceUsage: true,
		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
			k8sConfig = config.GetConfigOrDie()
			clientSet = kubernetes.NewForConfigOrDie(k8sConfig)
			rookClientSet = rookcli.NewForConfigOrDie(k8sConfig)
		},
		RunE: func(cmd *cobra.Command, args []string) (err error) {
			// See that here we can do what is requited to be done
			defaultClassProvisioner,err:= findDefaultClass()
			if err != nil {
				return err
			}
			output,err:= runCheck(defaultClassProvisioner);
			if err != nil {
				return err
			}

			// OUTPUT the result
			fmt.Sprintf("the availble free disk space with thr provisioner %s is: (%s)",
				output,defaultClassProvisioner.Name)
			return nil
		},
	}

	cmd.Flags().StringVar(&forStorageClass, "storageclass", "", "The storage class for which to check the free disk space.")
	cmd.Flags().StringVar(&openEBSImage, "openebs-image", "eeacms/rsync:2.3", "The image used by OpenEBS disk free evaluation pod.")
	cmd.Flags().StringVar(&openEBSNode, "openebs-node", "", "Show OpenEBS free disk space only for provided node.")
	return cmd
}

func runCheck(*v1.StorageClass) (string, err){
	switch selectedClass.Provisioner {
	case openEBSLocalProvisioner:
		....

	default:
		return fmt.Errorf("provisioner %q is not supported", selectedClass.Provisioner)
	}
}

func findDefaultClass() (*v1.StorageClass, error) {
	var selectedClass *v1.StorageClass
	classes, err := clientSet.StorageV1().StorageClasses().List(cmd.Context(), metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to list cluster storage classes: %w", err)
	}

	for _, class := range classes.Items {
		val, ok := class.Annotations[isDefaultStorageClassAnnotation]
		if ok && val == "true" && forStorageClass == "" {
			selectedClass = &class
			logger.Printf("Selected default storage class %q", selectedClass.Name)
			break
		}

		if class.Name == forStorageClass {
			selectedClass = &class
			logger.Printf("Selected storage class %q", selectedClass.Name)
			break
		}
	}

	if selectedClass == nil {
		if forStorageClass == "" {
			return fmt.Errorf("failed to find default storage class")
		}
		return fmt.Errorf("failed to find storage class %q", forStorageClass)
	}
	return selectedClass,
}
```

WDYT? Is that make sense?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:125]: If we change the version would still be that the default value?
Could we create a const on top and set the default value with a comment on top for help in the future if we need to understand it?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:86]: @ricardomaraschini see that in this implementation we will

print value 
print value
error found and returned / stop

Example:

$ /tmp/kurl cluster free-disk-space --storageclass openebs
62421426176     ip-172-31-60-177
62894407680     ip-172-31-80-103
Error 

Therefore, I think we should change that for return the error only after check all values OR just return the output when errors are not found. WDYT is that make sense?

Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:86]: This output won't happen. Or we print the output or the error. Why do you think it is possible ?

Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:90]: HI @ricardomaraschini, 

> This output won't happen. Or we print the output or the error. Why do you think it is possible ?

We have a loop and for each time we print the value see:
> fmt.Printf("%d\t%s\n", volume.Free, node) 
Then, if an error occurs in the loop we return an error and we stop to check.

So, how the scenario above will not happens?
To do what you are saying here `Or we print the output or the error` we can not use fmt.Print in the loop. We need to store the values in array and just print them after all process.`

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:78]: fmt.Print will print before that fails right?

Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:78]: Yes, this is the log of the freeSpaceGetter and it makes sense to print it if something goes wrong.

Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:78]: OK . You told that the idea is print the value that we have + the errors so I think we can close this one.
Thank you for the clarifications.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:90]: I see here we will continue. Sorry, I did not look this one properly.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:44]: If we want to not show the usage when we did not use the flags accordingly or an error occurs that is fine 👍

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:23]: How we will keep it maintained?
Will we need to update the default value when pvmigrate change it there?
Can we just use it from there or at least add a TODO for we export this const there and get here?
Is that something that make sense?

Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:23]: We don't need to keep them both in sync, no. There is a flag that can be used to replace the image as well. What else to you think is necessary ? If you prefer I can simply remove the default and let then customise in the installer.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:44]: That is fine, it is in use in other commands of the same binary.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:23]: I think the default value is great and it probably is the latest release one.
Yes, i see that we can use the flag but would be great if we do not need update that manually when a new release for example happens. However, that is not a blocker for we get this one merged.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:46]: The suggestion look for cohesion so we can grow in the future. 
This method has more than 30 lines which would not pass in a lint check for complexity for example. 

Anyway, I will not block the PR because of this.

**Review by camilamacedo86 — DISMISSED** [review]

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:131]: @ricardomaraschini here, how can we inform the node? Is an IP, is an label? Can we add an example?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:43]:
```suggestion
		Example:      "kurl cluster check-free-disk-space --storageclass openebs",
```

Can we improve this example by adding all possible flags with values? So that we know how to use it?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:129]:
```suggestion
	cmd.Flags().StringVar(&forStorageClass, "storageclass", "", fmt.Sprintf("Inform the storage class for which to check the free disk space. [Options: %s, %s, %s]", openEBSLocalProvisioner, rookRBDProvisioner, rookCephFSProvisioner ))
```

	isDefaultStorageClassAnnotation = "storageclass.kubernetes.io/is-default-class"
	openEBSLocalProvisioner         = "openebs.io/local"
	rookRBDProvisioner              = "rook-ceph.rbd.csi.ceph.com"
	rookCephFSProvisioner           = "rook-ceph.cephfs.csi.ceph.com"

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:130]:
```suggestion
	cmd.Flags().StringVar(&openEBSImage, "openebs-image", defaultOpenEBSPodImage, fmt.Sprintf("The image used by OpenEBS disk free evaluation pod. If not informed the default image used is %s", defaultOpenEBSPodImage))
```

**Review by camilamacedo86 — COMMENTED** [review]
I added just a few nits for we improve the flags desc and examples.
Otherwise, 
/lgtm

**Review by camilamacedo86 — APPROVED** [review]
Approved 🥇 

But see if you agree to address the nits for we improve the examples and flags description.
Please, feel free to ping me out to approve again as well.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:131]: Made the node flag clear (it is by node's name). Added to the sample. Described the supported provisioners.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:43]: Done.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:46]: I agree with adding complexity when it is necessary and I don't consider that this is the case. Moreover 30 lines for go (where possibly half would be error treatment) is not a long function.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:23]: As this is not a blocker I am going to resolve it.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:46]: As this is not a blocker I am resolving it.

**Review by camilamacedo86 — DISMISSED** [review]

File: pkg/cli/cluster_free_disk_space.go
Comment (emosbaugh) [inline pkg/cli/cluster_free_disk_space.go:35]: what do you think about using cmd.ErrOrStderr() here?

File: pkg/cli/cluster_free_disk_space.go
Comment (emosbaugh) [inline pkg/cli/cluster_free_disk_space.go:38]: can you instantiate these in the run() function?

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [reply pkg/cli/cluster_free_disk_space.go:38]: IHMO it should to be PreRunE since is a pre-requirement as suggested https://github.com/replicatedhq/kURL/pull/3784#discussion_r1034843469

I know that the methods here will exit(1) if fails but I also think that should be inside of the command so that if errors occurs we cobra would get that. ( i.e it is within silenceUsage is true but if want to remove that, cobra would output the usage/help for the command as well )

Comment (emosbaugh) [reply pkg/cli/cluster_free_disk_space.go:38]: im fine with prerune. i just dont like that to get the help text you have to be able to instantiate the kube client

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:219]: Why do we need this one?
Do we really a command that checks if the value is > than an a value ? 
Can not this check be done where this info is useful?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:205]: I do not think that we need this implementation here.
If we will use the check in a place where is required to know if the amount is or not bigger than X then this code can be there. I have a command that check the free space why this command also will check if the free space is bigger than a value informed? It seems out of the scope. Could we please remove this one?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:175]: Same of above can we use a const for the text here?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:205]:
```suggestion
```
I do not think that we need this implementation here.
If we will use the check in a place where is required to know if the amount is or not bigger than X then this code can be there. I have a command that check the free space why this command also will check if the free space is bigger than a value informed? It seems out of the scope. Could we please remove this one?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:228]:
```suggestion
```
Why do we need this one?
Do we really a command that checks if the value is > than an a value ? 
Can not this check be done where this info is useful?

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:193]: We are just checking if the value is or not supported above.
Ideally we do like:
a) First check all inputs ( return an error if something fails )
b) After we check the inputs we run the command. In this case we can return an error here before to check if the values are or nor supported that seems a little strange.

File: pkg/cli/cluster_free_disk_space.go
Comment (camilamacedo86) [inline pkg/cli/cluster_free_disk_space.go:74]:
```suggestion
```

Why we would add that in the check? 
What is the motivation?

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go]: @emosbaugh can you take a look at the new version and see if this still applies?

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:165]: No. Feel free to open a PR to address this once this is merged.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:175]: Same answer.

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go:74]: See my previous answers

**Review by emosbaugh — DISMISSED** [review]

**Review by camilamacedo86 — APPROVED** [review]

File: pkg/cli/cluster_free_disk_space.go
Comment (ricardomaraschini) [reply pkg/cli/cluster_free_disk_space.go]: @emosbaugh I have used the `debug` flag to enable the output to stderr. Can you take a look ?

**Review by camilamacedo86 — APPROVED** [review]

**Pull request merged**
